### ISSUE TITLE
Multiple PVC claims at similar time hang in Pending state for excessive time

### ISSUE THREAD

**Issue by tosseth** [IBM/ibm-spectrum-scale-csi#428]
**Describe the bug**
When multiple PVC claims are being requested (for example 5 PVCs), the PVCs can remain in the Pending state for an excessive amount of time (greater than 20 minutes). If a single PVC claim is made, it seems to go quickly enough (0-3 minutes).

Note: This bug seems to be a serialization problem with REST API calls from CSI to the Spectrum Scale cluster for set quota (REST cmd "quotas"), where multiple set quotas take forever to hash out. There is an issue opened against the Spectrum Scale team for this issue. This CSI bug issue is being opened in coordination with that defect, for tracking purposes.

**To Reproduce**
1. Create 5 PVC yamls for `kind: PersistentVolumeClaim`. I used 100G in size. (Not sure if this larger size makes an impact)
2. `oc apply -f <yamls>` in a loop, to force claims around the same timeframe.
3. Watch `oc get pvc` for the PVCs. They stay in Pending for a long time, until going to Bound state.

**Expected behavior**
The PVCs should bind quickly.

**Environment**
Please run the following an paste your output here:
``` 
10:18:25 [cpst-ocp-cluster-c:zen-c-automated] root@arcx3650fxxnh cpst-ocp-cluster-c 1 # oc -n ibm-spectrum-scale-ns exec ibm-spectrum-scale-core-bpb4p -- rpm -qa | grep gpfs
gpfs.gskit-8.0.55-12.x86_64
gpfs.crypto-5.1.0-3.x86_64
gpfs.msg.en_US-5.1.0-3.noarch
gpfs.adv-5.1.0-3.x86_64
gpfs.gui.cloud.deps-5.1.0-3.210301.111426.noarch
gpfs.compression-5.1.0-3.x86_64
gpfs.gpl-5.1.0-3.noarch
gpfs.docs-5.1.0-3.noarch
gpfs.gss.pmcollector-5.1.0-3.el8.x86_64
gpfs.base-5.1.0-3.x86_64
gpfs.license.dm-5.1.0-3.x86_64
gpfs.gss.pmsensors-5.1.0-3.el8.x86_64

10:18:37 [cpst-ocp-cluster-c:zen-c-automated] root@arcx3650fxxnh cpst-ocp-cluster-c # kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"7eec1ed62a6d1df3da17e9cf43782141093829a7", GitTreeState:"clean", BuildDate:"2021-04-22T19:35:31Z", GoVersion:"go1.15.7", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0+d856161", GitCommit:"d8561614ec39e7dc33ef531690b919178010226c", GitTreeState:"clean", BuildDate:"2021-04-23T13:59:46Z", GoVersion:"go1.15.7", Compiler:"gc", Platform:"linux/amd64"}
10:18:48 [cpst-ocp-cluster-c:zen-c-automated] root@arcx3650fxxnh cpst-ocp-cluster-c #

10:17:30 [cpst-ocp-cluster-c:zen-c-automated] root@arcx3650fxxnh cpst-ocp-cluster-c # oc get pods -n ibm-spectrum-scale-ns
NAME                                           READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-core-bpb4p                  1/1     Running   0          22d
ibm-spectrum-scale-core-jb2fc                  1/1     Running   0          22d
ibm-spectrum-scale-core-vbmnj                  1/1     Running   0          22d
ibm-spectrum-scale-gui-0                       9/9     Running   0          7d20h
ibm-spectrum-scale-operator-66c979bf67-fzqt8   1/1     Running   0          8d
ibm-spectrum-scale-pmcollector-0               2/2     Running   0          8d
ibm-spectrum-scale-pmcollector-1               2/2     Running   0          8d
10:17:50 [cpst-ocp-cluster-c:zen-c-automated] root@arcx3650fxxnh cpst-ocp-cluster-c #

10:18:48 [cpst-ocp-cluster-c:zen-c-automated] root@arcx3650fxxnh cpst-ocp-cluster-c # oc get pods -n ibm-spectrum-scale-csi-driver
NAME                                               READY   STATUS    RESTARTS   AGE
ibm-spectrum-scale-csi-26w4f                       2/2     Running   0          7d20h
ibm-spectrum-scale-csi-8czgh                       2/2     Running   0          7d20h
ibm-spectrum-scale-csi-attacher-0                  1/1     Running   0          8d
ibm-spectrum-scale-csi-jtsgc                       2/2     Running   0          9d
ibm-spectrum-scale-csi-operator-67448f6956-4s4sd   1/1     Running   0          8d
ibm-spectrum-scale-csi-provisioner-0               1/1     Running   0          8d
10:19:50 [cpst-ocp-cluster-c:zen-c-automated] root@arcx3650fxxnh cpst-ocp-cluster-c #
```
**Screenshots**

**Additional context**
Internal dev RTC tracking defect 275723

**Comment by tosseth** [comment]
The below shows the outstanding REST API jobs, from the Spectrum Scale storage cluster perspective. 
There are 2 jobs remaining in the "RUNNING" state for an excessive time. 
```
08:47:08 [cpst-ocp-cluster-c:default] root@arcx3650fxxnh ~ # curl -k -u 'xxx:xxx' 'https://<stg-gui-node>:443/scalemgmt/v2/jobs?fields=:all:' |grep -E "status|runtime|url" |tail -100
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 44455    0 44455    0     0   258k      0 --:--:-- --:--:-- --:--:--  258k
..
..
    "status" : "COMPLETED",
    "runtime" : 8342,
      "url" : "/scalemgmt/v2/filesystems/fsc/filesets"
    "status" : "COMPLETED",
    "runtime" : 600769,
      "url" : "/scalemgmt/v2/filesystems/fsc/quotas"
    "status" : "FAILED",
    "runtime" : 600001,
      "url" : "/scalemgmt/v2/filesystems/fsc/quotas"
    "status" : "COMPLETED",
    "runtime" : 605518,
      "url" : "/scalemgmt/v2/filesystems/fsc/quotas"
    "status" : "FAILED",
    "runtime" : 600005,
      "url" : "/scalemgmt/v2/filesystems/fsc/quotas"
    "status" : "FAILED",
    "runtime" : 600008,
      "url" : "/scalemgmt/v2/filesystems/fsc/quotas"
    "status" : "COMPLETED",
    "runtime" : 468,
      "url" : "/scalemgmt/v2/filesystems/fsc/directory/pvc-fc324700-fd97-40f5-9ab4-b2d08bea8c62%2Fpvc-fc324700-fd97-40f5-9ab4-b2d08bea8c62-data"
    "status" : "COMPLETED",
    "runtime" : 468,
      "url" : "/scalemgmt/v2/filesystems/fsc/directory/pvc-48e04915-9ee2-4802-800f-bd434a34eb6f%2Fpvc-48e04915-9ee2-4802-800f-bd434a34eb6f-data"
    "status" : "COMPLETED",
    "runtime" : 462,
      "url" : "/scalemgmt/v2/filesystems/fsc/quotas"
    "status" : "RUNNING",
    "runtime" : 553878,
      "url" : "/scalemgmt/v2/filesystems/fsc/quotas"
    "status" : "RUNNING",
    "runtime" : 553503,
      "url" : "/scalemgmt/v2/filesystems/fsc/quotas"
    "status" : "COMPLETED",
    "runtime" : 485,
      "url" : "/scalemgmt/v2/filesystems/fsc/directory/pvc-ace4d0ec-c9c6-46da-8a01-ab5de0e91c9d%2Fpvc-ace4d0ec-c9c6-46da-8a01-ab5de0e91c9d-data"
  "status" : {
08:51:35 [cpst-ocp-cluster-c:default] root@arcx3650fxxnh ~ #
```

Those correlate to these 2 PVCs remaining in the "Pending" state for quite a long time.
```
08:32:45 [cpst-ocp-cluster-c:default] root@arcx3650fxxnh cpst-ocp-cluster-c # while (( 1 ));do echo --- $(date) ---;oc get pvc |grep Pending;sleep 30;done
--- Thu Jun 17 08:33:01 MST 2021 ---
ibm-spectrum-scale-sc-todd-100g-2-pvc    Pending                                      ibm-spectrum-scale-sc   70s
ibm-spectrum-scale-sc-todd-100g-3-pvc    Pending                                      ibm-spectrum-scale-sc   70s
ibm-spectrum-scale-sc-todd-100g-4-pvc    Pending                                      ibm-spectrum-scale-sc   70s
ibm-spectrum-scale-sc-todd-100g-5-pvc    Pending                                      ibm-spectrum-scale-sc   69s
ibm-spectrum-scale-sc-todd-100gi-1-pvc   Pending                                      ibm-spectrum-scale-sc   71s
..
..
--- Thu Jun 17 08:53:38 MST 2021 ---
ibm-spectrum-scale-sc-todd-100g-4-pvc    Pending                                                                        ibm-spectrum-scale-sc   21m
ibm-spectrum-scale-sc-todd-100g-5-pvc    Pending                                                                        ibm-spectrum-scale-sc   21m
```

**Comment by deeghuge** [comment]
Hi @tosseth , Does you filesystem has `--perfileset-quota` on you filesystem? 
you can check it using `mmlsfs fsc --perfileset-quota`

**Comment by tosseth** [comment]
The perfileset-quota is not on. 
```
[root@stg-node0 fsc]# mmlsfs fsc --perfileset-quota
flag                value                    description
------------------- ------------------------ -----------------------------------
 --perfileset-quota No                       Per-fileset quota enforcement
[root@stg-node0 fsc]#
```

When setting up CSI deployment, I followed the steps from this procedure, which instruct to not set perfileset-quota.
https://www.ibm.com/docs/en/scalecontainernative?topic=driver-performing-pre-installation-tasks-csi-deployment

**Comment by tosseth** [comment]
Update for this issue and resolution/fix in Spectrum Scale.

**Summary:**

When multiple PVCs are created at roughly the same time, it can sometimes cause the IBM Spectrum Scale storage cluster handling the REST API calls from the IBM Spectrum Scale CSI driver to hang.
This hang occurs during the set quota calls made for each fileset that is backing the PVC. The REST calls are eventually completed and the PVCs are Bound successfully, but can take several minutes or even over an hour.

This can impact the installation of some assemblies for CloudPak for Data, such as Common Core Services, which creates multiple PVCs at the same time during the install. The long hang for the PVCs to complete can cause the install to timeout and fail.

There is no known workaround (other than continue to retry the assembly install until it succeeds).
However a fix for the issue has been created and verified by the CPST test team. The fix for the issue is targeted for an upcoming Spectrum Scale release 5.1.1.x (internal tracker RTC 275723).

**Comment by deeghuge** [comment]
Closing since issue is fixed in Scale GUI version => 5.1.1.2
